src/gshhs/gshhs.1
.TH GSHHS 1 "15 Jul 2011" "GMT 4.5.7" "Generic Mapping Tools"
.SH NAME
gshhs \- Extract ASCII listings from binary GSHHS or WDBII data files
.SH SYNOPSIS
\fBgshhs\fP \fIbinaryfile.b\fP [ \fB\-A\fP\fIarea\fP ] [ \fB\-G\fP ] [ \fB\-I\fP\fIid\fP ] [ \fB\-L\fP ] [ \fB\-M\fP ] 
[ \fB\-N\fP\fIlevel\fP ] [ \fB\-Qe\fP|\fBi\fP ] > \fIasciifile.txt\fP
.SH DESCRIPTION
\fBgshhs\fP reads the binary coastline (GSHHS) or political boundary or river (WDBII) files and
extracts an ASCII listing.  It automatically handles byte-swabbing between different
architectures.  Optionally, only segment header info can be displayed.  The output header information has
the format \fIID npoints  hierarchical-level source area f_area west east south north container ancestor\fP,
where hierarchical levels for coastline polygons go from 1 (shoreline) to 4 (lake inside
island inside lake inside land).  Source is either W (World Vector Shoreline) or C (CIA World
Data Bank II); lower case is used if a lake is a river-lake (a portion of a river that is so wide it is
better represented by a closed polygon). The \fIwest east south north\fP is
the enclosing rectangle, \fIarea\fP is the polygon area in km^2 while f_area is the actual area of the
ancestor polygon (at full resolution), \fIcontainer\fP is the ID of the polygon that contains this polygon (-1 if none),
and \fIancestor\fP is the ID of the polygon in the full resolution set that was reduced to yield this
polygon (-1 if full resolution since there is no ancestor).
For river and border data the header is simply
\fIID npoints  hierarchical-level source west east south north\fP.  
For more information about the file formats, see TECHNICAL INFORMATION below.
.TP
\fIbinaryfile.b\fP
GSHHS or WDBII binary data file as distributed with the GSHHS data supplement.  Any of the
5 standard resolutions (full, high, intermediate, low, crude) can be used.
.TP
\fB\-A\fP
Only output polygons whose area equals or exceeds the \fIarea\fP value in km^2 [Default outputs all polygons].
.TP
\fB\-G\fP
Write output that can be imported into GNU Octave or Matlab by ending each segment with a NaN-record.
.TP
\fB\-I\fP
Only output information for the polygon that matches \fIid\fP.
Use \fB\-Ic\fP to get all the continents only [Default outputs all polygons].
.TP
\fB\-L\fP
Only output a listing of polygon or line segment headers [Default outputs headers and data records].
.TP
\fB\-M\fP
Start all header records with the GMT multiple segment indicator '>' [Default uses P for polygons
and L for lines].
.TP
\fB\-N\fP
Only output features whose level matches the given \fIlevel\fP [Default will output all levels].
.TP
\fB\-Q\fP
Control what to do with river-lakes (river sections large enough to be stored as closed polygons).
Use \fB\-Qe\fP to exclude them and \fB\-Qi\fP to exclude everything else instead [Default outputs all polygons].
.SH EXAMPLES
.sp
To convert the entire intermediate GSHHS binary data to ASCII files for Octave/Mathlab, run
.br
.sp
gshhs gshhs_i.b -G > gshhs_i.txt
.br
.sp
To only get a listing of the headers for the river data set at full resolution, try
.br
.sp
gshhs wdb_rivers_f.b -L > riverlisting.txt
.br
.sp
To only extract lakes, excluding river-lakes, from the high resolution file, try
.br
.sp
gshhs gshhs_h.b -Ee -N2 > all_lakes.txt
.SH TECHNICAL INFORMATION
Users who wish to access the GSHHS or WDBII data directly from their custom programs should
consult the gshhs.c and gshhs.h source code and familiarize themselves with the data format and
how various information flags are packed into a single 4-byte integer.  While we do not maintain
any Octave/Matlab code to read these files we are aware that both Mathworks and IDL have made
such tools available to their users.  However, they tend not to update their code and our file
structure has evolved considerably over time, breaking their code.  Here, some general technical
comments on the binary data files are given.
.br
\fBGSHHS\fP: These files contain completely closed polygons of continents and islands (level 1),
lakes (level 2), islands-in-lakes (level 3) and ponds-in-islands-in-lakes (level 4); a particular
level can be extracted using the \fB\-N\fP option.  Continents are identified as the first 6 polygons
and can be extracted via the \fB\-Ic\fP option.  The IDs for the continents are Eurasia (0),
Africa (1), North America (2), South America (3), Antarctica (4), and Australia (5).
Files are sorted on area from large to small.  There are two sub-groups for level 2: Regular lakes and the
so-called "river-lakes", the latter being sections of a river that are so wide to warrant a polygon
representation.  These river-lakes are flagged in the header (also see \fB\-Q\fP).  All five resolutions
are free of self-intersections.  Areas of all features have been computed using a Lambert azimuthal
equal-area projection centered on the polygon centroids, using WGS-84 as the ellipsoid.  GMT use the
GSHHS as a starting point but then partition the polygons into pieces using a resolution-dependent
binning system; parts of the world are then rebuilt into closed polygons on the fly as needed.
For more information on GSHHS processing, see Wessel and Smith (1996).
.br
\fBWDBII\fP.  These files contain sets of line segments not necessarily in any particular order.  Thus,
it is not possible to extract information pertaining to just one river or one country.  Furthermore,
the 4 lower resolutions derive directly from the full resolution by application of the Douglas-Peucker
algorithm (see gshhs_dp), hence self-intersections are increasingly likely as the resolution is
degraded.  Note that the river-lakes included in GSHHS are also duplicated in the WDBII river files
so that each data set can be a stand-alone representation.  Users who wish to access both data sets
can recognize the river-lakes features by examining the header structure (see the source code for
details); they are also the only closed polygons in the WDBII river file.  There are many levels (classes) in the
river file: River-lakes (0), Permanent major rivers (1), Additional major rivers (2), Additional rivers (3),
Minor rivers (4), Intermittent rivers -- major (6), Intermittent rivers -- additional (7),
Intermittent rivers -- minor (8), Major canals (10), Canals of lesser importance (11), and
Canals -- irrigation type (12).  For the border file there are three levels: National boundaries (1),
Internal domestic boundaries (2), and international maritime boundaries (3).  Individual levels or classes
may be extracted via \fB\-N\fP.
.SH REFERENCES
Douglas, D. H., and T. K. Peucker, 1973, Algorithms for the reduction
of the number of points required to represent a digitized line
of its caricature, \fICan. Cartogr., 10\fP, 112-122.
.br
Gorny, A. J., 1977, \fIWorld Data Bank II General User GuideRep. PB 271869\fP, 10pp,
Central Intelligence Agency, Washington, DC.
.br
Soluri, E. A., and V. A. Woodson, 1990, World Vector Shoreline, \fIInt. Hydrograph. Rev., LXVII(1)\fP, 27-35.
.br
Wessel, P., and W. H. F. Smith, 1996, A global, self-consistent, hierarchical, high-resolution
shoreline database, \fIJ. Geophys. Res., 101(B4)\fP, 8741-8743.
.SH "SEE ALSO"
.IR GMT (1),
.IR gshhs_dp (1)
.IR gshhstograss (1)


ChangeLog
-----------------------------------------------------------------------------------------
	$Id: ChangeLog,v 1.944 2011/07/14 20:50:58 guru Exp $
-----------------------------------------------------------------------------------------
-----------------------------------------------------------------------------------------
2011-07-15	pwessel
	* GMT 4.5.7 finalized and released
-----------------------------------------------------------------------------------------

2011-07-14	pwessel
	* pslib.c:		The @-, @+, @# would reset size to original font size, ignoring
				any changes a @:<size>: sequence would have made.  Also, when
				@:<size>: was used the derived sizes for sub/super were not updated.

2011-07-13	remko
	* gmt_plot.c:		Force arcs to be counter-clockwise in GMT_pie and GMT_matharc.

2011-07-11	pwessel
	* install_gmt.sh:	Simplified GMT packaging to one tarball, with all of GSHHS in
				a separate tarball [similar for the Windows installers].

2011-07-11	remko
	* doc/GNUmakefile:	tex4ht now writes src="file" instead of src=file, so had to
				adjust sed script.
	* website:		webexamples now only in 100dpi form (from doc/html/images).
	* Makefile:		Updated doc directory.

2011-07-08	pwessel
	* *.c:			Added initialization of pointers to NULL as in GMT 5.
	* makecpt.c:		Ignore any files given on command line (issue warning).

2011-07-08	remko
	* doc/GNUmakefile:	Fixed a few lingering transition issues.

2011-07-05	pwessel
	* gmt_support.c:	Used hardwired DEG_TO_KM with Earth's mean radius meant wrong
				results for other planets.  Now using the ELLIPSOID values.

2011-06-30	remko
	* share/doc/gmt/examples:	Moved to doc

2011-06-23	pwessel
	* gmt_shore.c:		Now reads new GSHHS 2.2.0 format that includes feature areas
				stored as doubles.  The old int(0.1*area) would set all
				areas < 0.1 km^2 to zero [only affected gshhs supplement].
				Areas were also recomputed to use WGS-84 ellipsoid (not sphere).
	* gshhs/gshhs.c:	Updated to handle GSHHS 2.2.0.	Also added -Amin to exclude
				smaller polygons on output and -G for special Octave/Matlab-
				formatted output [suggested by José Luis García Pallero],
				as well as -Q to control output of river-lakes and -N to
				control which levels to get.

2011-06-20	pwessel
	* *.c:			No longer use DIR_DELIM since / works for all.  Added path
				fixer to turn odd /c/dir/.. into c:/dir/..

2011-06-20	remko
	* gmt_notunix.h:	Only define PATH_SEPARATOR (as char), not also PATH_DELIM.

2011-06-15	jluis
	* makecpt:		For some crazy reason on Windows the comment line now was 
				being written at the end of the cpt file.

2011-06-13	pwessel
	* -X -Y:		Allow absolute positioning with -Ya even if no -X is given
				and vice versa.

2011-06-12	pwessel
	* x2sys:		New test suite based on Wessel [2010] figure scripts.

2011-06-10	pwessel
	* gmt_map.c:		Clip path for van Grinten was wrong for global maps.
	
2011-06-08	pwessel
	* xgrid/xgrid_Panner.c:	Eliminate "cast to pointer from integer of different size"
				by casting the value via (ptrdiff_t) [Thanks to L. Parkes].

2011-06-07	pwessel
	* animations:		Add note that when run without any argument they produce just
				the first frame as PS.  With an arg we build the movie.
	* grdcontour.c:		Could crash if -C -A give no contours and we freed a non-allocated array.
				[Thanx to Walter Harms].

2011-06-05	pwessel
	* nearneighbor.c:	For global region in longitudes, cannot extend w/e limits
				as we do for Cartesian and region-limited areas.

2011-06-01	pwessel
	* spotter/grdrotater.c:	Could end up with w=e=0 in some cases.
	* grdedit.c:		The -N option did not consider replicate w/e points for
				gridline-registered grids.
	* grdfilter.c:		Now gives consistent values at poles and along shared E/W cols.

2011-05-31	pwessel
	* grdfilter.c:		Convolution weights were y-transposed so gave wrong results
				at/near the poles for spherical filtering.  Also, some nodes
				were duplicated in the convolution, resulting in inconsistent
				values at the S or N pole.
	* grdlandmask.c:	Did not correctly truncate nodes for GSHHS bins with no data.

2011-05-26	pwessel
	* gmt_map.c:		GMT_wesn_clip needed to adjust longitudes to fit the given domain.

2011-05-25	pwessel
	* nearneighbor.c:	The -N1 selection did not reset the minimum sector setting to 1.

2011-05-23	jluis
	* xyz2grd.c:		Fixed bug when reading a file using -E option on Windows. 
				The file was actually read with GMT_fopen and accessed with fgets 

2011-05-11	remko
	* gmt_nc.c:		Ignore reading of node_offset attribute. Registration is now
				based on match between actual_range attributes and coordinate arrays.

2011-05-10	pwessel
	* gmt_grdio.c:		Did not account for the doubling of a grid array for complex
				data when scaling data after read.  This made grdfft give
				odd results when grids with a scale other than 1 was read.
				Bug was first introduced in GMT 3.0 in 1995, making it a
				bug with seniority!

2011-05-09	pwessel
	* grdimage.c:		When image is grayscale and -Q is used then image must be
				converted to 24-bit and we set NaN color to a non-gray value.

2011-05-07	pwessel
	* gmt_grdio.c:		Failed to create proper old-style v3 netcdf file if selected
				for the output in grdblend.
	* gmt_agc_io.c:		Needed char floatvalue[PARAMSIZE+1], with the 1 to hold final \0
	* grdblend.c:		Now mentions which formats are supported and polices the process.

2011-05-03	pwessel
	* pstext.c:		The -A can reset the text angle but also needs to reset justification.

2011-05-03	remko
	* gmt_map.c, pstext.c:	Now allow negative azimuth in -E option.

2011-05-01	remko
	* script/*.sh:		Replaced ". functions.sh" by ". ./functions.sh" to avoid problems
				with people who do not have . in their PATH. E.g., Ubuntu.

2011-05-01	pwessel
	* job03.sh:		Replaced complicated awk scripting with gmtmath to ensure
				we get the common sampling range which changed by a tad
				due to the bug in project.
2011-04-28	pwessel

src/gshhs/gshhs.c
/*	$Id: gshhs.c,v 1.51 2011/07/08 20:17:10 guru Exp $
 *
 *	Copyright (c) 1996-2011 by P. Wessel and W. H. F. Smith
 *	See LICENSE.TXT file for copying and redistribution conditions.
 *
 * PROGRAM:	gshhs.c
 * AUTHOR:	Paul Wessel (pwessel@hawaii.edu)
 * CREATED:	JAN. 28, 1996
 * PURPOSE:	To extract ASCII data from the binary GSHHS shoreline data
 *		as described in the 1996 Wessel & Smith JGR Data Analysis Note.
 * VERSION:	1.1 (Byte flipping added)
 *		1.2 18-MAY-1999:
 *		   Explicit binary open for DOS systems
 *		   POSIX.1 compliant
 *		1.3 08-NOV-1999: Released under GNU GPL
 *		1.4 05-SEPT-2000: Made a GMT supplement; FLIP no longer needed
 *		1.5 14-SEPT-2004: Updated to deal with latest GSHHS database (1.3)
 *		1.6 02-MAY-2006: Updated to deal with latest GSHHS database (1.4)
 *		1.7 11-NOV-2006: Fixed bug in computing level (&& vs &)
 *		1.8 31-MAR-2007: Updated to deal with latest GSHHS database (1.5)
 *		1.9 27-AUG-2007: Handle line data as well as polygon data
 *		1.10 15-FEB-2008: Updated to deal with latest GSHHS database (1.6)
 *		1.11 15-JUN-2009: Now contains information on container polygon,
 *				the polygons ancestor in the full resolution, and
 *				a flag to tell if a lake is a riverlake.
 *				Updated to deal with latest GSHHS database (2.0)
 *		1.12 24-MAY-2010: Deal with 2.1 format.
 *		1.13 15-JUL-2011: Now contains improved area information (2.2.0),
 *				 and revised greenwhich flags (now 2-bit; see gshhs.h).
 *				 Also added -A and -G as suggested by José Luis García Pallero,
 *				 as well as -Qe|i to control river-lake output, and -N to
 *				 get a particular level.
 *
 *	This program is free software; you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation; version 2 or any later version.
 *
 *	This program is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *
 *	Contact info: www.soest.hawaii.edu/pwessel */

#include "gshhs.h"

int main (int argc, char **argv)
{
	double w, e, s, n, area, f_area, lon, lat, scale = 10.0, minarea = 0.0;
	char source, kind[2] = {'P', 'L'}, c = '>', *file = NULL;
	char *name[2] = {"polygon", "line"}, container[8], ancestor[8];
	FILE *fp = NULL;
	int k, line, max_east = 270000000, info, single, error, ID, flip, m, octave = 0, rank = -1;
	int  OK, level, version, greenwich, river, src, msformat = 0, qmode = 0, first = 1;
	size_t n_read;
	struct GSHHS h;
 	struct POINT p;
       
	info = single = error = ID = 0;
	for (k = 1; k < argc; k++) {
		if (argv[k][0] == '-') {	/* Option switch */
			switch (argv[k][1]) {
				case 'A':
					minarea = atof (&argv[k][2]);
					break;
				case 'G':
					octave = 1;
					break;
				case 'L':
					info = 1;
					break;
				case 'M':
					msformat = 1;
					break;
				case 'N':
					rank = atoi (&argv[k][2]);
					break;
				case 'I':
					if (argv[k][2] == 'c')
						single = 2;
					else {
						single = 1;
						ID = atoi (&argv[k][2]);
					}
					break;
				case 'Q':
					if (argv[k][2] == 'e')
						qmode = 1;
					else if (argv[k][2] == 'i')
						qmode = 2;
					else
						error++;
					break;
				default:
					error++;
					break;
			}
		}
		else
			file = argv[k];
	}

	if (!file) error++;
	
	if (argc == 1 || error) {
		fprintf (stderr, "gshhs %s - ASCII export of GSHHS/WDBII %s data\n\n", GSHHS_PROG_VERSION, GSHHS_DATA_VERSION);
		fprintf (stderr, "usage:  gshhs gshhs|wdbii_rivers|wdb_borders_[f|h|i|l|c].b [-A<area>] [-G] [-I<id>] [-L] [-M] [-N<level>] [-Qe|i] > ascii.txt\n\n");
		fprintf (stderr, "\t-A will extract only polygons whose area is greater or equal than <area>.\n");
		fprintf (stderr, "\t-G will write '%%' at start of each segment header [P or L] (overwrites -M)\n"
				 "\t   and write 'NaN NaN' after each segment to enable import by GNU Octave or Matlab.\n");
		fprintf (stderr, "\t-I Output data for polygon number <id>.  Use -Ic to get all continent polygons\n");
		fprintf (stderr, "\t   [Default is all polygons].\n");
		fprintf (stderr, "\t-L will only list headers (no data output).\n");
		fprintf (stderr, "\t-M will write '>' at start of each segment header [P or L].\n");
		fprintf (stderr, "\t-N will only output features whose level matches <level> [Default is all levels].\n");
		fprintf (stderr, "\t-Q controls river-lakes.  By default all polygons are output.\n");
		fprintf (stderr, "\t   use -Qe to exclude river-lakes, and -Qi to ONLY get river-lakes.\n");
		exit (EXIT_FAILURE);
	}
	if (!file) {
		fprintf (stderr, "gshhs: No data file given!\n");
		error++;
	}

	if ((fp = fopen (file, "rb")) == NULL ) {
		fprintf (stderr, "gshhs:  Could not find file %s.\n", file);
		exit (EXIT_FAILURE);
	}
		
	n_read = fread ((void *)&h, (size_t)sizeof (struct GSHHS), (size_t)1, fp);
	version = (h.flag >> 8) & 255;
	flip = (version != GSHHS_DATA_RELEASE);	/* Take as sign that byte-swabbing is needed */
	
	while (n_read == 1) {
		if (flip) {
			h.id = swabi4 ((unsigned int)h.id);
			h.n  = swabi4 ((unsigned int)h.n);
			h.west  = swabi4 ((unsigned int)h.west);
			h.east  = swabi4 ((unsigned int)h.east);
			h.south = swabi4 ((unsigned int)h.south);
			h.north = swabi4 ((unsigned int)h.north);
			h.area  = swabi4 ((unsigned int)h.area);
			h.area_full  = swabi4 ((unsigned int)h.area_full);
			h.flag  = swabi4 ((unsigned int)h.flag);
			h.container = swabi4 ((unsigned int)h.container);
			h.ancestor  = swabi4 ((unsigned int)h.ancestor);
		}
		level = h.flag & 255;				/* Level is 1-4 */
		version = (h.flag >> 8) & 255;			/* Version is 1-255 */
		if (first) fprintf (stderr, "gshhs %s - Found GSHHS/WDBII version %d in file %s\n", GSHHS_PROG_VERSION, version, file);
		greenwich = (h.flag >> 16) & 3;			/* Greenwich is 0-3 */
		src = (h.flag >> 24) & 1;			/* Greenwich is 0 (WDBII) or 1 (WVS) */
		river = (h.flag >> 25) & 1;			/* River is 0 (not river) or 1 (is river) */
		w = h.west  * GSHHS_SCL;			/* Convert from microdegrees to degrees */
		e = h.east  * GSHHS_SCL;
		s = h.south * GSHHS_SCL;
		n = h.north * GSHHS_SCL;
		source = (src == 1) ? 'W' : 'C';		/* Either WVS or CIA (WDBII) pedigree */
		if (river) source = tolower ((int)source);	/* Lower case c means river-lake */
		line = (h.area) ? 0 : 1;			/* Either Polygon (0) or Line (1) (if no area) */
		if (version >= 9) {				/* Variable magnitude for area scale */
			m = h.flag >> 26;
			scale = pow (10.0, (double)m);		/* Area scale */
		}
		area = h.area / scale;				/* Now im km^2 */
		f_area = h.area_full / scale;			/* Now im km^2 */

		OK = ((!single || ((single == 1 && h.id == ID) || (single == 2 && h.id <= 5))) && area >= minarea);	/* Skip if not the one (-I) or too small (-A) */
		if (OK && qmode && ((river && qmode == 1) || (!river && qmode == 2))) OK = 0;	/* Skip if riverlake/not riverlake (-Q) */
		if (OK && rank >= 0 && level != rank) OK = 0;		/* Skip if not the right level (-N) */
		first = 0;
		
		if (!msformat) c = kind[line];
	        if (octave) c = '%';
		if (OK) {
			if (line)
				printf ("%c %6d%8d%3d%2c%10.5f%10.5f%10.5f%10.5f\n", c, h.id, h.n, level, source, w, e, s, n);
			else {
				(h.container == -1) ? sprintf (container, "-") : sprintf (container, "%6d", h.container);
				(h.ancestor == -1) ? sprintf (ancestor, "-") : sprintf (ancestor, "%6d", h.ancestor);
				printf ("%c %6d%8d%2d%2c %.12g %.12g%11.5f%11.5f%10.5f%10.5f %s %s\n", c, h.id, h.n, level, source, area, f_area, w, e, s, n, container, ancestor);
			}
		}

		if (info || !OK) {	/* Skip data, only want headers */
			fseek (fp, (long)(h.n * sizeof(struct POINT)), SEEK_CUR);
		}
		else {
			for (k = 0; k < h.n; k++) {

				if (fread ((void *)&p, (size_t)sizeof(struct POINT), (size_t)1, fp) != 1) {
					fprintf (stderr, "gshhs:  Error reading file %s for %s %d, point %d.\n", argv[1], name[line], h.id, k);
					exit (EXIT_FAILURE);
				}
				if (flip) {
					p.x = swabi4 ((unsigned int)p.x);
					p.y = swabi4 ((unsigned int)p.y);
				}
				lon = p.x * GSHHS_SCL;
				if (((greenwich & 1) && p.x > max_east) || (h.west > 180000000)) lon -= 360.0;
				lat = p.y * GSHHS_SCL;
				printf ("%11.6f%11.6f\n", lon, lat);
			}
			if (octave) printf ("NaN\tNaN\n");
		}
		max_east = 180000000;	/* Only Eurasia needs 270 */
		n_read = fread((void *)&h, (size_t)sizeof (struct GSHHS), (size_t)1, fp);
	}
		
	fclose (fp);

	exit (EXIT_SUCCESS);
}
